Update maven detections

AyanSinhaMahapatra opened the pull request:
Update maven package and license detections:

* Fix MANIFEST.MF parsing
* Add package assembly for both pom.xml and MANIFEST.MF
* implement functionality to do package specific license detection
* run license detection on whole extracted license statement for maven
* Add rules for license detection fixes
* Fix detection for pom.properties file
* Fix --todo plugin for package detection
* Fix false positive case for bare word rules
* Fix issue of unknown license detection in package manifest

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁
* [x] Looked for possible updates in documentation and added updates if applicable
* [x] Updated CHANGELOG.rst

AyanSinhaMahapatra committed:
Fix parsing of MANIFEST.MF files

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add maven license detection updates

Add a new config variable for datafile handlers, which if enabled
will run a package-ecosystem specific implementation of license
detection. Added a function to implement this for maven which only keeps
the relevant information and runs license detection on the whole
license statement and not on the respective values/attributes
one-by-one. Enabled this for maven to improve license detection there.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/packagedcode/maven.py:
```diff
@@ -1171,6 +1172,63 @@ def _parse(
+        
+        for license_entry in new_extracted_license:
+            license_entry.pop("distribution")
+            if not license_entry.get("name", None):
```
Nit: None is already the default for a get

pombredanne replied:
Also what about this instead:
`{key: value for key, value in license_entry.items() if value and key != 'distribution'}`

AyanSinhaMahapatra replied:
Done, thanks!

AyanSinhaMahapatra replied:
> Also what about this instead:
> {key: value for key, value in license_entry.items() if value and key != 'distribution'}

Actually we are doing a `pop()` so we actually not have empty key/value pairs in the original extracted license statement. Which is correct since we are creating these key/value pairs irrespective of them being present in the first place.

JonoYang commented on src/packagedcode/jar_manifest.py:
```diff
@@ -278,7 +296,7 @@ def dget(s):
 
     # Bundle-DocURL: http://logging.apache.org/log4j/1.2
-    package['documentation_url'] = dget('Bundle-DocURL')
+    #package['documentation_url'] = dget('Bundle-DocURL')
```
Should this line be removed if it's being commented out?

AyanSinhaMahapatra replied:
Adding these in the `extra_data` field instead, thanks.

JonoYang commented on src/packagedcode/jar_manifest.py:
```diff
@@ -369,7 +387,7 @@ def dget(s):
     # Require-Bundle
 
-    package['notes'] = dget('Comment')
+    #package['notes'] = dget('Comment')
```
See comment on line 299

AyanSinhaMahapatra replied:
Addressed, thanks!

AyanSinhaMahapatra committed:
Fix maven package assembly

Adds a fix for maven packacge assembly, combining MANIFEST.MF and
pom.xml files into a single package, instead of multiple top level
packages. Also fixes pom.properties file parsing.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Refactor and address feedback

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

JonoYang commented on src/packagedcode/jar_manifest.py:
```diff
@@ -369,10 +339,61 @@ def dget(s):
+    from packagedcode.maven import JavaJarManifestHandler
+    from packagedcode.maven import JavaOSGiManifestHandler
+
+    if package_type == 'maven':
```
```suggestion
    if package_type in ['maven', 'jar']:
        return JavaJarManifestHandler.datasource_id
```

Minor change just to avoid repetition.

AyanSinhaMahapatra replied:
Added this, thanks!

JonoYang reviewed (changes requested):
@AyanSinhaMahapatra Looks good, I just highlighted some small issues.

AyanSinhaMahapatra commented on tests/summarycode/data/classify/with_package_data.expected.json:
```diff
@@ -500,15 +542,64 @@
+          "source_packages": [],
+          "file_references": [],
+          "extra_data": {
+            "pom_properties": {
```
Aren't these the namespace/name/version? Why are we having these in the extra-data instead of having these in the proper attributes (and then using these in the assembly)?

AyanSinhaMahapatra replied:
I'm adding these in the main package attributes instead, so we have the correct PURL for these

AyanSinhaMahapatra committed:
Fix misc license/package detection bugs

* Add rules for license detection fixes
* Fix detection for pom.properties file
* Fix --todo plugin for package detection
* Fix false positive case for bare word rules
* Fix issue of unknown license detection in package manifest

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add more license rules

Adds additional license rules to take care of maven
package license detection adjustments.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented on src/licensedcode/data/rules/bsd-3-clause-no-nuclear-warranty_and_jj2000_2.RULE:
```diff
@@ -0,0 +1,10 @@
+
+name: {{JJ2000}}
+url: LICENSE-JJ2000.txt
+name: {{BSD 3-clause License w/nuclear disclaimer}}
```
could it better to split these in two rules?

AyanSinhaMahapatra replied:
Done, thanks!

pombredanne commented on src/licensedcode/data/rules/json_17.RULE:
```diff
@@ -0,0 +1,32 @@
+---
+license_expression: json
+is_license_tag: yes
```
```suggestion
is_license_text: yes
```

AyanSinhaMahapatra replied:
Done, thanks!

pombredanne commented on src/licensedcode/data/rules/lgpl-2.1_and_apache-2.0_1.RULE:
```diff
@@ -0,0 +1,12 @@
+
+name: LGPL, version 2.1
+url: http://www.gnu.org/licenses/licenses.html
+name: Apache License v2.0
```
Would it be better to split this in two?

AyanSinhaMahapatra replied:
Done, thanks!

pombredanne commented on src/licensedcode/data/rules/lgpl-2.1_and_apache-2.0_2.RULE:
```diff
@@ -0,0 +1,12 @@
+
+name: LGPL, version 2.1
+url: http://www.gnu.org/licenses/licenses.html
+name: ASL, version 2
```
Would it be better to split this in two?

AyanSinhaMahapatra replied:
Done, thanks!

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -729,6 +749,27 @@ def collect_license_detections(codebase, include_license_clues=True):
 
+
+def verify_package_license_expression(license_detection_mappings, license_expression):
+
```
What is this method doing and returning? Some doc would help.

AyanSinhaMahapatra replied:
Docstring added, thanks!

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -978,9 +1019,13 @@ def is_false_positive(license_matches, package_license=False):
+    bare_rules = ['gpl_bare', 'freeware_bare', 'public-domain_bare']
+    is_bare_rule = all(
+        any([
+            True
```
If you use this, prefer this code .... but IMHO we should have a flag on the Rule model instead
```suggestion
            bare_rule in license_match.rule.identifier
```

AyanSinhaMahapatra replied:
Yeah this could be a computed flag in the Rule object allright, will implement this seperately. Thanks!
For now adding the suggested change.

pombredanne commented on src/licensedcode/detection.py:
```diff
@@ -978,9 +1019,13 @@ def is_false_positive(license_matches, package_license=False):
+        any([
+            True
+            for bare_rule in bare_rules
+            if bare_rule in license_match.rule.identifier
```
```suggestion
```

AyanSinhaMahapatra replied:
Added this, thanks!

pombredanne commented on src/packagedcode/__init__.py:
```diff
@@ -253,3 +253,12 @@ def get_package_handler(package_data):
+    maven.MavenPackageData
+]
+
+PACKAGE_DATA_CLASS_BY_DATASOURCE_ID = {
```
Create the data at once statically rather that doing this work on each import
```suggestion
PACKAGE_DATA_CLASS_BY_DATASOURCE_ID = {maven.MavenPackageData.datasource_id: maven.MavenPackageData}
```

AyanSinhaMahapatra replied:
This would be better, doing this instead, thanks!

pombredanne commented on src/packagedcode/jar_manifest.py:
```diff
@@ -369,10 +339,67 @@ def dget(s):
+    if comment or doc_url:
+        package["extra_data"] = {}
+        if comment:
+            package["extra_data"]['notes'] = comment
```
we need to find a better way than using extra_data.

AyanSinhaMahapatra replied:
See https://github.com/nexB/scancode-toolkit/pull/3447#discussion_r1262231378. We might want to add fields to the PackageData class then. And we might handle this separately, AFAIK this data is not really used anywhere and this is just a fix to make sure MANIFEST parsing returns everything else, and not crash.

pombredanne commented on src/packagedcode/jar_manifest.py:
```diff
@@ -278,7 +248,7 @@ def dget(s):
 
     # Bundle-DocURL: http://logging.apache.org/log4j/1.2
-    package['documentation_url'] = dget('Bundle-DocURL')
+    doc_url = dget('Bundle-DocURL')
```
I get what you are doing, but we are losing data here.

AyanSinhaMahapatra replied:
We are returning this `package` as-is to be used for creating a PackageData class, see https://github.com/nexB/scancode-toolkit/blob/update-maven-detections/src/packagedcode/maven.py#L138 with `models.PackageData(**manifest,)`. We don't have a `documentation_url` or a `notes` field in PackageData, and this is why the Manifest parsing used to fail and not return anything. I'm just ensuring we return whatever we can, instead of crashing and not returning anything. And since I'm having this in `extra_data`, we're not really losing any data IMHO

pombredanne commented on src/packagedcode/jar_manifest.py:
```diff
@@ -369,10 +339,67 @@ def dget(s):
     return package
 
 
+def parse_scm_connection(scm_connection):
```
@TG1999 ping ... is this related or relevant to your purldb work?

pombredanne commented on src/packagedcode/jar_manifest.py:
```diff
@@ -369,10 +339,67 @@ def dget(s):
+    if package_type in ['maven', 'jar']:
+        return JavaJarManifestHandler.datasource_id
+    elif package_type == 'osgi':
+        return JavaOSGiManifestHandler.datasource_id
```
What if a maven package also has OSGi thing?s

AyanSinhaMahapatra replied:
We can improve the parsing and handle more cases allright, but let's do this separately through another issue, and if you have an example of that case please provide the same. But this was also a fix, and I'm only making sure this works without crashing here.

pombredanne commented on src/packagedcode/licensing.py:
```diff
@@ -764,22 +767,33 @@ def get_license_detections_and_expression(
-        expression_symbols=expression_symbols,
-    )
+    if datasource_id in PACKAGE_DATA_CLASS_BY_DATASOURCE_ID:
+        package_data_class = PACKAGE_DATA_CLASS_BY_DATASOURCE_ID.get(datasource_id, None)
```
```suggestion
        package_data_class = PACKAGE_DATA_CLASS_BY_DATASOURCE_ID.get(datasource_id)
```

AyanSinhaMahapatra replied:
Done, thanks!

pombredanne reviewed (changes requested):
Please see some nits for your consideration

AyanSinhaMahapatra commented:
@pombredanne @JonoYang addressed all your comments, thanks!

pombredanne commented on src/licensedcode/data/rules/bsd-3-clause-no-nuclear-warranty_12.RULE:
```diff
@@ -0,0 +1,7 @@
+---
+license_expression: bsd-3-clause-no-nuclear-warranty
+is_license_tag: yes
```
You could add a:

```suggestion
is_license_tag: yes
referenced_filenames:
    - LICENSE-Sun.txt
```

AyanSinhaMahapatra replied:
Added this thanks!

pombredanne commented on src/licensedcode/data/rules/jj2000_1.RULE:
```diff
@@ -0,0 +1,7 @@
+---
+
+name: {{JJ2000}}
+url: LICENSE-JJ2000.txt
```
You could add a referenced filenames for this

AyanSinhaMahapatra replied:
Added this thanks!

pombredanne commented on src/packagedcode/licensing.py:
```diff
@@ -764,22 +767,33 @@ def get_license_detections_and_expression(
-        approximate=approximate,
-        expression_symbols=expression_symbols,
-    )
+    if datasource_id in PACKAGE_DATA_CLASS_BY_DATASOURCE_ID:
```
```suggestion
    package_data_class = PACKAGE_DATA_CLASS_BY_DATASOURCE_ID.get(datasource_id)
    if package_data_class:
```

AyanSinhaMahapatra replied:
Added this, thanks!

pombredanne commented on src/packagedcode/licensing.py:
```diff
@@ -764,22 +767,33 @@ def get_license_detections_and_expression(
-        expression_symbols=expression_symbols,
-    )
+    if datasource_id in PACKAGE_DATA_CLASS_BY_DATASOURCE_ID:
+        package_data_class = PACKAGE_DATA_CLASS_BY_DATASOURCE_ID.get(datasource_id)
```
```suggestion

```

AyanSinhaMahapatra replied:
Added this, thanks!

pombredanne reviewed (approved):
LGTM with a few extra nitrs that you can fix or not as you like before merging. Thanks!
